// @supabase/ui/dist/esm/components/Menu/Menu2.js
import { jsx, jsxs } from 'react/jsx-runtime';
import Typography from '../Typography/Typography.js';
import { MenuContextProvider, useMenuContext } from './MenuContext.js';
import styleHandler from '../../lib/theme/styleHandler.js';

function Menu({
  children,
  className,
  style,
  type = 'text'
}) {
  return jsx("nav", Object.assign({
    role: "menu",
    "aria-label": "Sidebar",
    "aria-orientation": "vertical",
    "aria-labelledby": "options-menu",
    className: className,
    style: style
  }, {
    children: jsx(MenuContextProvider, Object.assign({
      type: type
    }, {
      children: jsx("ul", {
        children: children
      }, void 0)
    }), void 0)
  }), void 0);
}

function Item({
  children,
  icon,
  active,
  rounded,
  onClick,
  doNotCloseOverlay = false,
  showActiveBar = false,
  style
}) {
  const __styles = styleHandler('menu');

  const {
    type
  } = useMenuContext();
  let classes = [__styles.item.base];
  classes.push(__styles.item.variants[type].base);

  if (active) {
    classes.push(__styles.item.variants[type].active);
  } else {
    classes.push(__styles.item.variants[type].normal);
  }

  let contentClasses = [__styles.item.content.base];

  if (active) {
    contentClasses.push(__styles.item.content.active);
  } else {
    contentClasses.push(__styles.item.content.normal);
  }

  let iconClasses = [__styles.item.icon.base];

  if (active) {
    iconClasses.push(__styles.item.icon.active);
  } else {
    iconClasses.push(__styles.item.icon.normal);
  }

  return jsx("li", Object.assign({
    role: "menuitem",
    className: "outline-none"
  }, {
    children: jsxs("a", Object.assign({
      style: style,
      className: classes.join(' '),
      onClick: onClick,
      "aria-current": active ? 'page' : undefined
    }, {
      children: [icon && jsx("span", Object.assign({
        className: iconClasses.join(' ')
      }, {
        children: icon
      }), void 0), jsx("span", Object.assign({
        className: contentClasses.join(' ')
      }, {
        children: children
      }), void 0)]
    }), void 0)
  }), void 0);
}
function Group({
  children,
  icon,
  title
}) {
  const __styles = styleHandler('menu');

  const {
    type
  } = useMenuContext();
  return jsxs("div", Object.assign({
    className: [__styles.group.base, __styles.group.variants[type]].join(' ')
  }, {
    children: [icon && jsx("span", Object.assign({
      className: __styles.group.icon
    }, {
      children: icon
    }), void 0), jsx("span", Object.assign({
      className: __styles.group.content
    }, {
      children: title
    }), void 0), children]
  }), void 0);
}
function Misc({
  children
}) {
  return jsx("div", {
    children: jsx(Typography.Text, {
      children: jsx("span", {
        children: children
      }, void 0)
    }, void 0)
  }, void 0);
}
Menu.Item = Item;
Menu.Group = Group;
Menu.Misc = Misc;

export { Group, Item, Misc, Menu as default };
