// @supabase/ui/dist/cjs/components/Menu/Menu2.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var Typography = require('../Typography/Typography.js');
var MenuContext = require('./MenuContext.js');
var styleHandler = require('../../lib/theme/styleHandler.js');

function Menu({
  children,
  className,
  style,
  type = 'text'
}) {
  return jsxRuntime.jsx("nav", Object.assign({
    role: "menu",
    "aria-label": "Sidebar",
    "aria-orientation": "vertical",
    "aria-labelledby": "options-menu",
    className: className,
    style: style
  }, {
    children: jsxRuntime.jsx(MenuContext.MenuContextProvider, Object.assign({
      type: type
    }, {
      children: jsxRuntime.jsx("ul", {
        children: children
      }, void 0)
    }), void 0)
  }), void 0);
}

function Item({
  children,
  icon,
  active,
  rounded,
  onClick,
  doNotCloseOverlay = false,
  showActiveBar = false,
  style
}) {
  const __styles = styleHandler['default']('menu');

  const {
    type
  } = MenuContext.useMenuContext();
  let classes = [__styles.item.base];
  classes.push(__styles.item.variants[type].base);

  if (active) {
    classes.push(__styles.item.variants[type].active);
  } else {
    classes.push(__styles.item.variants[type].normal);
  }

  let contentClasses = [__styles.item.content.base];

  if (active) {
    contentClasses.push(__styles.item.content.active);
  } else {
    contentClasses.push(__styles.item.content.normal);
  }

  let iconClasses = [__styles.item.icon.base];

  if (active) {
    iconClasses.push(__styles.item.icon.active);
  } else {
    iconClasses.push(__styles.item.icon.normal);
  }

  return jsxRuntime.jsx("li", Object.assign({
    role: "menuitem",
    className: "outline-none"
  }, {
    children: jsxRuntime.jsxs("a", Object.assign({
      style: style,
      className: classes.join(' '),
      onClick: onClick,
      "aria-current": active ? 'page' : undefined
    }, {
      children: [icon && jsxRuntime.jsx("span", Object.assign({
        className: iconClasses.join(' ')
      }, {
        children: icon
      }), void 0), jsxRuntime.jsx("span", Object.assign({
        className: contentClasses.join(' ')
      }, {
        children: children
      }), void 0)]
    }), void 0)
  }), void 0);
}
function Group({
  children,
  icon,
  title
}) {
  const __styles = styleHandler['default']('menu');

  const {
    type
  } = MenuContext.useMenuContext();
  return jsxRuntime.jsxs("div", Object.assign({
    className: [__styles.group.base, __styles.group.variants[type]].join(' ')
  }, {
    children: [icon && jsxRuntime.jsx("span", Object.assign({
      className: __styles.group.icon
    }, {
      children: icon
    }), void 0), jsxRuntime.jsx("span", Object.assign({
      className: __styles.group.content
    }, {
      children: title
    }), void 0), children]
  }), void 0);
}
function Misc({
  children
}) {
  return jsxRuntime.jsx("div", {
    children: jsxRuntime.jsx(Typography['default'].Text, {
      children: jsxRuntime.jsx("span", {
        children: children
      }, void 0)
    }, void 0)
  }, void 0);
}
Menu.Item = Item;
Menu.Group = Group;
Menu.Misc = Misc;

exports.Group = Group;
exports.Item = Item;
exports.Misc = Misc;
exports['default'] = Menu;
