// @supabase/ui/dist/esm/components/InputNumber/InputNumber.js
import { jsx, jsxs } from 'react/jsx-runtime';
import React__default, { useEffect } from 'react';
import { FormLayout } from '../../lib/Layout/FormLayout/FormLayout.js';
import InputErrorIcon from '../../lib/Layout/InputErrorIcon.js';
import InputIconContainer from '../../lib/Layout/InputIconContainer.js';
import InputNumberStyles from './InputNumber.module.css.js';
import { useFormContext } from '../Form/FormContext.js';
import Space from '../Space/Space2.js';

function InputNumber({
  autoComplete,
  autofocus,
  className,
  defaultValue,
  descriptionText,
  disabled,
  error,
  icon,
  id = '',
  inputRef,
  label,
  afterLabel,
  beforeLabel,
  labelOptional,
  layout,
  name = '',
  onChange,
  onBlur,
  onFocus,
  onKeyDown,
  placeholder,
  value = undefined,
  style,
  size = 'medium',
  min,
  max,
  borderless = false,
  validation
}) {
  const {
    formContextOnChange,
    values,
    errors,
    handleBlur,
    touched,
    fieldLevelValidation
  } = useFormContext();
  if (values && !value) value = values[id];
  if (handleBlur) onBlur = handleBlur;

  if (!error) {
    if (errors && !error) error = errors[id || name];
    error = touched && touched[id || name] ? error : undefined;
  }

  function onInputChange(e) {
    if (onChange) onChange(e); // update form

    if (formContextOnChange) formContextOnChange(e); // run field level validation

    if (validation) fieldLevelValidation(id, validation(e.target.value));
  }

  useEffect(() => {
    if (validation) fieldLevelValidation(id, validation(value));
  }, []);
  const inputClasses = [InputNumberStyles['sbui-inputnumber']];
  const inputRefCurrent = inputRef ? inputRef : /*#__PURE__*/React__default.createRef();
  if (error) inputClasses.push(InputNumberStyles['sbui-inputnumber--error']);
  if (icon) inputClasses.push(InputNumberStyles['sbui-inputnumber--with-icon']);

  if (size) {
    inputClasses.push(InputNumberStyles[`sbui-inputnumber--${size}`]);
  }

  if (borderless) inputClasses.push(InputNumberStyles['sbui-inputnumber--borderless']);

  return jsx("div", Object.assign({
    className: className
  }, {
    children: jsx(FormLayout, Object.assign({
      label: label,
      afterLabel: afterLabel,
      beforeLabel: beforeLabel,
      labelOptional: labelOptional,
      layout: layout,
      id: id,
      error: error,
      descriptionText: descriptionText,
      style: style,
      size: size
    }, {
      children: jsxs("div", Object.assign({
        className: InputNumberStyles['sbui-inputnumber-container']
      }, {
        children: [jsx("input", {
          autoComplete: autoComplete,
          autoFocus: autofocus,
          defaultValue: defaultValue,
          disabled: disabled,
          id: id,
          name: name,
          onChange: onInputChange,
          onFocus: onFocus ? event => onFocus(event) : undefined,
          onBlur: onBlur,
          onKeyDown: onKeyDown ? event => onKeyDown(event) : undefined,
          placeholder: placeholder,
          ref: inputRefCurrent,
          type: 'number',
          value: value,
          className: inputClasses.join(' '),
          min: min,
          max: max
        }, void 0), icon && jsx(InputIconContainer, {
          icon: icon
        }, void 0), error ? jsx(Space, Object.assign({
          className: InputNumberStyles['sbui-inputnumber-actions-container'],
          size: 1
        }, {
          children: error && jsx(InputErrorIcon, {
            size: size
          }, void 0)
        }), void 0) : null]
      }), void 0)
    }), void 0)
  }), void 0);
}

export { InputNumber as default };
