// @supabase/ui/dist/cjs/components/Tabs/Tabs2.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var TabsContext = require('./TabsContext.js');
var TabsPrimitive = require('@radix-ui/react-tabs');
var styleHandler = require('../../lib/theme/styleHandler.js');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () {
            return e[k];
          }
        });
      }
    });
  }
  n['default'] = e;
  return Object.freeze(n);
}

var TabsPrimitive__namespace = /*#__PURE__*/_interopNamespace(TabsPrimitive);

function Tabs({
  children,
  defaultActiveId,
  activeId,
  type = 'pills',
  size = 'tiny',
  block,
  onChange,
  onClick,
  scrollable,
  addOnBefore,
  addOnAfter,
  listClassNames
}) {
  const [activeTab, setActiveTab] = React.useState(defaultActiveId ? defaultActiveId : // if no defaultActiveId is set use the first panel
  children && children[0].props ? children[0].props.id : '');

  let __styles = styleHandler['default']('tabs'); // activeId state can be overriden externally with `active`
  // defaults to use activeTab


  const active = activeId ? activeId : activeTab;

  function onTabClick(id) {
    const newTabSelected = id !== active;
    setActiveTab(id);
    if (onClick) onClick(id);
    if (onChange && newTabSelected) onChange(id);
  } // for styling the tabs for underline style

  const listClasses = [__styles[type].list];
  if (scrollable) listClasses.push(__styles.scrollable);
  if (listClassNames) listClasses.push(listClassNames); // if only 1 react child, it needs to be converted to a list/array
  // this is so 1 tab can be displayed

  if (children && !Array.isArray(children)) {
    children = [children];
  }

  return jsxRuntime.jsxs(TabsPrimitive__namespace.Root, Object.assign({
    defaultValue: defaultActiveId,
    value: active,
    className: __styles.base
  }, {
    children: [jsxRuntime.jsxs(TabsPrimitive__namespace.List, Object.assign({
      className: listClasses.join(' ')
    }, {
      children: [addOnBefore, children.map(tab => {
        const activeMatch = active === tab.props.id;
        const triggerClasses = [__styles[type].base, __styles.size[size]];

        if (activeMatch) {
          triggerClasses.push(__styles[type].active);
        } else {
          triggerClasses.push(__styles[type].inactive);
        }

        if (block) {
          triggerClasses.push(__styles.block);
        }

        return jsxRuntime.jsxs(TabsPrimitive__namespace.Trigger, Object.assign({
          onKeyDown: e => {
            if (e.keyCode === 13) {
              e.preventDefault();
              onTabClick(tab.props.id);
            }
          },
          onClick: () => onTabClick(tab.props.id),
          value: tab.props.id,
          className: triggerClasses.join(' ')
        }, {
          children: [tab.props.icon, jsxRuntime.jsx("span", {
            children: tab.props.label
          }, void 0)]
        }), `${tab.props.id}-tab-button`);
      }), addOnAfter]
    }), void 0), jsxRuntime.jsx(TabsContext.TabsContext.Provider, Object.assign({
      value: {
        activeId: active
      }
    }, {
      children: children
    }), void 0)]
  }), void 0);
}

function Panel({
  children,
  id
}) {
  let __styles = styleHandler['default']('tabs');

  return jsxRuntime.jsx(TabsContext.TabsContext.Consumer, {
    children: ({
      activeId
    }) => {
      return jsxRuntime.jsx(TabsPrimitive__namespace.Content, Object.assign({
        value: id,
        className: __styles.content
      }, {
        children: children
      }), void 0);
    }
  }, void 0);
}
Tabs.Panel = Panel;

exports.Panel = Panel;
exports['default'] = Tabs;
