// @supabase/ui/dist/cjs/components/SidePanel/SidePanel2.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../_virtual/_tslib.js');
var jsxRuntime = require('react/jsx-runtime');
var Dialog = require('@radix-ui/react-dialog');
var styleHandler = require('../../lib/theme/styleHandler.js');
var Button = require('../Button/Button2.js');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () {
            return e[k];
          }
        });
      }
    });
  }
  n['default'] = e;
  return Object.freeze(n);
}

var Dialog__namespace = /*#__PURE__*/_interopNamespace(Dialog);

const SidePanel = _a => {
  var {
    id,
    className,
    children,
    header,
    visible,
    open,
    size = 'medium',
    loading,
    align = 'right',
    hideFooter = false,
    customFooter = undefined,
    onConfirm,
    onCancel,
    confirmText = 'Confirm',
    cancelText = 'Cancel',
    triggerElement,
    defaultOpen
  } = _a,
      props = _tslib.__rest(_a, ["id", "className", "children", "header", "visible", "open", "size", "loading", "align", "hideFooter", "customFooter", "onConfirm", "onCancel", "confirmText", "cancelText", "triggerElement", "defaultOpen"]);

  const __styles = styleHandler['default']('sidepanel'); // function stopPropagation(e: React.MouseEvent) {
  //   e.stopPropagation()
  // }


  const footerContent = customFooter ? customFooter : jsxRuntime.jsxs("div", Object.assign({
    className: __styles.footer
  }, {
    children: [jsxRuntime.jsx(Button['default'], Object.assign({
      disabled: loading,
      type: "default",
      onClick: () => onCancel ? onCancel() : null
    }, {
      children: cancelText
    }), void 0), jsxRuntime.jsx(Button['default'], Object.assign({
      loading: loading,
      onClick: () => onConfirm ? onConfirm() : null
    }, {
      children: confirmText
    }), void 0)]
  }), void 0);

  function handleOpenChange(open) {
    if (visible !== undefined && !open) {
      // controlled component behaviour
      if (onCancel) onCancel();
    }
  }

  open = open || visible;
  return jsxRuntime.jsxs(Dialog__namespace.Root, Object.assign({
    open: open,
    onOpenChange: handleOpenChange,
    defaultOpen: defaultOpen
  }, {
    children: [triggerElement && jsxRuntime.jsx(Dialog__namespace.Trigger, Object.assign({
      className: __styles.trigger
    }, {
      children: triggerElement
    }), void 0), jsxRuntime.jsxs(Dialog__namespace.Portal, {
      children: [jsxRuntime.jsx(Dialog__namespace.Overlay, {
        className: __styles.overlay
      }, void 0), jsxRuntime.jsxs(Dialog__namespace.Content, Object.assign({
        className: [__styles.base, __styles.size[size], __styles.align[align], className && className].join(' '),
        onOpenAutoFocus: props.onOpenAutoFocus,
        onCloseAutoFocus: props.onCloseAutoFocus,
        onEscapeKeyDown: props.onEscapeKeyDown,
        onPointerDownOutside: props.onPointerDownOutside,
        onInteractOutside: props.onInteractOutside
      }, {
        children: [header && jsxRuntime.jsx("header", Object.assign({
          className: __styles.header
        }, {
          children: header
        }), void 0), jsxRuntime.jsx("div", Object.assign({
          className: __styles.contents
        }, {
          children: children
        }), void 0), !hideFooter && footerContent]
      }), void 0)]
    }, void 0)]
  }), void 0);
};

function Seperator() {
  let __styles = styleHandler['default']('sidepanel');

  return jsxRuntime.jsx("div", {
    className: __styles.seperator
  }, void 0);
}
function Content({
  children
}) {
  let __styles = styleHandler['default']('sidepanel');

  return jsxRuntime.jsx("div", Object.assign({
    className: __styles.content
  }, {
    children: children
  }), void 0);
}
SidePanel.Content = Content;
SidePanel.Seperator = Seperator;

exports.Content = Content;
exports.Seperator = Seperator;
exports['default'] = SidePanel;
