// @supabase/postgres-meta/dist/main/sql/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.versionSql = exports.typesSql = exports.triggersSql = exports.tablesSql = exports.schemasSql = exports.rolesSql = exports.relationshipsSql = exports.publicationsSql = exports.primaryKeysSql = exports.policiesSql = exports.grantsSql = exports.functionsSql = exports.extensionsSql = exports.configSql = exports.columnsSql = void 0;
const fs_1 = require("fs");
const path_1 = require("path");
exports.columnsSql = fs_1.readFileSync(path_1.resolve(__dirname, 'columns.sql'), 'utf-8');
exports.configSql = fs_1.readFileSync(path_1.resolve(__dirname, 'config.sql'), 'utf-8');
exports.extensionsSql = fs_1.readFileSync(path_1.resolve(__dirname, 'extensions.sql'), 'utf-8');
exports.functionsSql = fs_1.readFileSync(path_1.resolve(__dirname, 'functions.sql'), 'utf-8');
exports.grantsSql = fs_1.readFileSync(path_1.resolve(__dirname, 'grants.sql'), 'utf-8');
exports.policiesSql = fs_1.readFileSync(path_1.resolve(__dirname, 'policies.sql'), 'utf-8');
exports.primaryKeysSql = fs_1.readFileSync(path_1.resolve(__dirname, 'primary_keys.sql'), 'utf-8');
exports.publicationsSql = fs_1.readFileSync(path_1.resolve(__dirname, 'publications.sql'), 'utf-8');
exports.relationshipsSql = fs_1.readFileSync(path_1.resolve(__dirname, 'relationships.sql'), 'utf-8');
exports.rolesSql = fs_1.readFileSync(path_1.resolve(__dirname, 'roles.sql'), 'utf-8');
exports.schemasSql = fs_1.readFileSync(path_1.resolve(__dirname, 'schemas.sql'), 'utf-8');
exports.tablesSql = fs_1.readFileSync(path_1.resolve(__dirname, 'tables.sql'), 'utf-8');
exports.triggersSql = fs_1.readFileSync(path_1.resolve(__dirname, 'triggers.sql'), 'utf-8');
exports.typesSql = fs_1.readFileSync(path_1.resolve(__dirname, 'types.sql'), 'utf-8');
exports.versionSql = fs_1.readFileSync(path_1.resolve(__dirname, 'version.sql'), 'utf-8');
