// @testdeck/jest/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.params = exports.skip = exports.only = exports.pending = exports.retries = exports.timeout = exports.slow = exports.test = exports.suite = void 0;
const core = require("@testdeck/core");
const jestRunner = {
    suite(name, callback, settings) {
        // TODO: Push settings.timeout on the stack and pass it to each child test...
        switch (settings && settings.execution) {
            case "only":
                describe.only(name, callback);
                break;
            case "skip":
                describe.skip(name, callback);
                break;
            case "pending":
                // No `describe(name);` nor `describe.todo`... Use skip.
                describe.skip(name, callback);
                break;
            default:
                describe(name, callback);
        }
    },
    test(name, callback, settings) {
        switch (settings && settings.execution) {
            case "only":
                it.only(name, callback, settings && settings.timeout);
                break;
            case "skip":
                it.skip(name, callback, settings && settings.timeout);
                break;
            case "pending":
                it.todo(name);
                break;
            default:
                it(name, callback, settings && settings.timeout);
        }
    },
    beforeAll(name, callback, settings) {
        beforeAll(callback, settings && settings.timeout);
    },
    beforeEach(name, callback, settings) {
        beforeEach(callback, settings && settings.timeout);
    },
    afterEach(name, callback, settings) {
        afterEach(callback, settings && settings.timeout);
    },
    afterAll(name, callback, settings) {
        afterAll(callback, settings && settings.timeout);
    }
};
class JestClassTestUI extends core.ClassTestUI {
    constructor(runner = jestRunner) {
        super(runner);
    }
}
const jestDecorators = new JestClassTestUI();
// "context" is not available for jest
exports.suite = jestDecorators.suite, exports.test = jestDecorators.test, exports.slow = jestDecorators.slow, exports.timeout = jestDecorators.timeout, exports.retries = jestDecorators.retries, exports.pending = jestDecorators.pending, exports.only = jestDecorators.only, exports.skip = jestDecorators.skip, exports.params = jestDecorators.params;
