// @supabase/ui/dist/cjs/components/Card/Card2.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var Typography = require('../Typography/Typography.js');
var styleHandler = require('../../lib/theme/styleHandler.js');

function Card({
  children,
  className,
  cover,
  hoverable,
  style,
  title,
  titleExtra
}) {
  let __styles = styleHandler['default']('card');

  let classes = [__styles.base];
  if (hoverable) classes.push(__styles.hoverable);
  if (className) classes.push(className);
  return jsxRuntime.jsxs("div", Object.assign({
    className: classes.join(' '),
    style: style
  }, {
    children: [title && jsxRuntime.jsxs("div", Object.assign({
      className: __styles.head
    }, {
      children: [jsxRuntime.jsx(Typography['default'].Text, Object.assign({
        style: {
          margin: 0
        }
      }, {
        children: title
      }), void 0), jsxRuntime.jsx(Typography['default'].Link, Object.assign({
        style: {
          margin: 0
        }
      }, {
        children: titleExtra
      }), void 0)]
    }), void 0), cover, jsxRuntime.jsx("div", Object.assign({
      className: __styles.content
    }, {
      children: children
    }), void 0)]
  }), void 0);
}

function Meta({
  title,
  description,
  style,
  className
}) {
  return jsxRuntime.jsxs("div", Object.assign({
    style: style,
    className: className
  }, {
    children: [jsxRuntime.jsx(Typography['default'].Title, Object.assign({
      style: {
        margin: '0'
      },
      level: 5
    }, {
      children: title
    }), void 0), jsxRuntime.jsx("div", {
      children: jsxRuntime.jsx(Typography['default'].Text, Object.assign({
        type: "secondary"
      }, {
        children: description
      }), void 0)
    }, void 0)]
  }), void 0);
}

Card.Meta = Meta;

exports['default'] = Card;
