// @supabase/ui/dist/cjs/components/ContextMenu/ContextMenu2.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var IconCheck = require('../Icon/icons/IconCheck/IconCheck2.js');
var RadixContextMenu = require('@radix-ui/react-context-menu');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () {
            return e[k];
          }
        });
      }
    });
  }
  n['default'] = e;
  return Object.freeze(n);
}

var RadixContextMenu__namespace = /*#__PURE__*/_interopNamespace(RadixContextMenu);

function ContextMenu({
  onOpenChange,
  alignOffset = 6,
  overlay,
  children,
  className,
  style
}) {
  // let classes = [ContextMenuStyles['sbui-contextmenu__content']]
  // if (className) {
  //   classes.push(className)
  // }
  return jsxRuntime.jsxs(RadixContextMenu__namespace.Root, Object.assign({
    onOpenChange: onOpenChange
  }, {
    children: [jsxRuntime.jsx(RadixContextMenu__namespace.Trigger // className={ContextMenuStyles['sbui-contextmenu__trigger']}
    , {
      children: children
    }, void 0), jsxRuntime.jsx(RadixContextMenu__namespace.Content, Object.assign({
      sideOffset: alignOffset,
      // className={classes.join(' ')}
      style: style
    }, {
      children: overlay
    }), void 0)]
  }), void 0);
}

function Item({
  children,
  icon,
  disabled,
  onClick
}) {
  return jsxRuntime.jsxs(RadixContextMenu__namespace.Item // className={ContextMenuStyles['sbui-contextmenu-item']}
  , Object.assign({
    // className={ContextMenuStyles['sbui-contextmenu-item']}
    disabled: disabled,
    onSelect: onClick
  }, {
    children: [icon && icon, jsxRuntime.jsx("span", {
      children: children
    }, void 0)]
  }), void 0);
}
function Misc({
  children,
  icon
}) {
  return jsxRuntime.jsxs("div", {
    children: [icon && icon, children]
  }, void 0);
}
function Checkbox({
  children,
  checked: propsChecked,
  onChange,
  disabled,
  ItemIndicator
}) {
  const [checked, setChecked] = React.useState(propsChecked ? propsChecked : false);

  const handleChange = e => {
    if (onChange) onChange(e);
    setChecked(e);
  };

  return jsxRuntime.jsxs(RadixContextMenu__namespace.CheckboxItem, Object.assign({
    checked: checked,
    onCheckedChange: handleChange,
    // className={`${ContextMenuStyles['sbui-contextmenu-item']} ${ContextMenuStyles['sbui-contextmenu-input']}`}
    disabled: disabled
  }, {
    children: [jsxRuntime.jsxs(RadixContextMenu__namespace.ItemIndicator // className={ContextMenuStyles['sbui-contextmenu-input__check']}
    , {
      children: [ItemIndicator ? ItemIndicator : jsxRuntime.jsx(IconCheck['default'], {
        size: "tiny"
      }, void 0), jsxRuntime.jsx(RadixContextMenu__namespace.CheckboxItem, {}, void 0)]
    }, void 0), jsxRuntime.jsx("span", {
      children: children
    }, void 0)]
  }), void 0);
}
function Radio({
  children,
  value,
  ItemIndicator
}) {
  return jsxRuntime.jsxs(RadixContextMenu__namespace.RadioItem, Object.assign({
    value: value
  }, {
    children: [jsxRuntime.jsx(RadixContextMenu__namespace.ItemIndicator // className={ContextMenuStyles['sbui-contextmenu-input__check']}
    , {
      children: ItemIndicator ? ItemIndicator : jsxRuntime.jsx(IconCheck['default'], {
        size: "tiny"
      }, void 0)
    }, void 0), jsxRuntime.jsx("span", {
      children: children
    }, void 0)]
  }), void 0);
}
function RadioGroup({
  children,
  value: propsValue,
  onChange
}) {
  const [value, setValue] = React.useState(propsValue ? propsValue : '');

  const handleChange = e => {
    if (onChange) onChange(e);
    setValue(e);
  };

  return jsxRuntime.jsx(RadixContextMenu__namespace.RadioGroup, Object.assign({
    value: value,
    onValueChange: handleChange
  }, {
    children: children
  }), void 0);
}
function Label({
  children
}) {
  return jsxRuntime.jsx(RadixContextMenu__namespace.Label // className={ContextMenuStyles['sbui-contextmenu-label']}
  , {
    children: children
  }, void 0);
}
ContextMenu.Item = Item;
ContextMenu.Misc = Misc;
ContextMenu.Checkbox = Checkbox;
ContextMenu.Radio = Radio;
ContextMenu.RadioGroup = RadioGroup;
ContextMenu.Label = Label;

exports.Checkbox = Checkbox;
exports.Item = Item;
exports.Label = Label;
exports.Misc = Misc;
exports.Radio = Radio;
exports.RadioGroup = RadioGroup;
exports['default'] = ContextMenu;
