// @supabase/ui/dist/esm/components/Form/Form2.js
import { __rest } from '../../_virtual/_tslib.js';
import { jsx } from 'react/jsx-runtime';
import { useReducer } from 'react';
import { useFormik } from 'formik';
import { FormContextProvider } from './FormContext.js';

function errorReducer(state, action) {
  if (!action.error) {
    const payload = state;
    delete payload[action.key];
    return payload;
  }

  if (action) {
    return Object.assign(Object.assign({}, state), {
      [action.key]: action.error
    });
  } else {
    throw new Error();
  }
}

function Form(_a) {
  var {
    validate
  } = _a,
      props = __rest(_a, ["validate"]);

  const [fieldLevelErrors, dispatchErrors] = useReducer(errorReducer, null);

  function handleFieldLevelValidation(key, error) {
    dispatchErrors({
      key,
      error
    });
  }

  const formik = useFormik(Object.assign(Object.assign({
    validateOnBlur: true
  }, props), {
    validationSchema: props.validationSchema,
    initialValues: props.initialValues,
    onSubmit: props.onSubmit,
    validate: validate || function () {
      return fieldLevelErrors;
    }
  })); // console.log('values', formik.values)
  // console.log('errors', formik.errors)
  // console.log('touched', formik.touched)

  return jsx("form", Object.assign({
    onSubmit: formik.handleSubmit
  }, {
    children: jsx(FormContextProvider, {
      values: formik.values,
      errors: formik.errors,
      formContextOnChange: formik.handleChange,
      handleBlur: formik.handleBlur,
      touched: formik.touched,
      fieldLevelValidation: handleFieldLevelValidation,
      // children={rest.children}
      children: props.children({
        /** map of field names to specific error for that field */
        errors: formik.errors,
        // /** map of field names to whether the field has been touched */
        touched: formik.touched,

        /** whether the form is currently submitting */
        isSubmitting: formik.isSubmitting,

        /** whether the form is currently validating (prior to submission) */
        isValidating: formik.isValidating,

        /** Number of times user tried to submit the form */
        submitCount: formik.submitCount
      })
    }, void 0)
  }), void 0);
}

export { Form as default };
