// @supabase/ui/dist/esm/components/Collapsible/Collapsible.js
import { __rest } from '../../_virtual/_tslib.js';
import { jsx } from 'react/jsx-runtime';
import * as RadixCollapsible from '@radix-ui/react-collapsible';
import styleHandler from '../../lib/theme/styleHandler.js';

const Collapsible = _a => {
  //   const [_open, setOpen] = React.useState(open)
  var {
    open = undefined,
    children,
    className
  } = _a,
      props = __rest(_a, ["open", "children", "className"]); // function handleOpenChange(e: boolean) {
  //   console.log(e)
  //   if (onChange) onChange(e)
  // }


  return jsx(RadixCollapsible.Root, Object.assign({
    defaultOpen: props.defaultOpen,
    open: open,
    onOpenChange: props.onOpenChange,
    disabled: props.disabled,
    className: className
  }, {
    children: children
  }), void 0);
};
function Trigger({
  children,
  asChild
}) {
  return jsx(RadixCollapsible.Trigger, Object.assign({
    asChild: asChild
  }, {
    children: children
  }), void 0);
}
function Content({
  children,
  className
}) {
  const __styles = styleHandler('collapsible');

  return jsx(RadixCollapsible.Content, Object.assign({
    className: [__styles.content, className].join(' ')
  }, {
    children: children
  }), void 0);
}
Collapsible.Trigger = Trigger;
Collapsible.Content = Content;

export { Collapsible, Content, Trigger, Collapsible as default };
