// @supabase/ui/dist/cjs/components/Popover/Popover.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var RadixPopover = require('@radix-ui/react-popover');
var IconX = require('../Icon/icons/IconX/IconX2.js');
var styleHandler = require('../../lib/theme/styleHandler.js');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () {
            return e[k];
          }
        });
      }
    });
  }
  n['default'] = e;
  return Object.freeze(n);
}

var RadixPopover__namespace = /*#__PURE__*/_interopNamespace(RadixPopover);

function Popover({
  align = 'center',
  ariaLabel,
  arrow = false,
  children,
  className,
  defaultOpen = false,
  modal,
  onOpenChange,
  open,
  overlay,
  portalled,
  side = 'bottom',
  sideOffset = 6,
  style,
  header,
  footer,
  size = 'content'
}) {
  const __styles = styleHandler['default']('popover');

  let classes = [__styles.content, __styles.size[size]];

  if (className) {
    classes.push(className);
  }

  return jsxRuntime.jsxs(RadixPopover__namespace.Root, Object.assign({
    defaultOpen: defaultOpen,
    modal: modal,
    onOpenChange: onOpenChange,
    open: open
  }, {
    children: [jsxRuntime.jsx(RadixPopover__namespace.Trigger // className={DropdownStyles['sbui-popover__trigger']}
    , Object.assign({
      // className={DropdownStyles['sbui-popover__trigger']}
      className: __styles.trigger,
      "aria-label": ariaLabel
    }, {
      children: children
    }), void 0), jsxRuntime.jsxs(RadixPopover__namespace.Content, Object.assign({
      sideOffset: sideOffset,
      side: side,
      align: align,
      className: classes.join(' '),
      style: style,
      portalled: portalled
    }, {
      children: [arrow && jsxRuntime.jsx(RadixPopover__namespace.Arrow // className={DropdownStyles['sbui-popover__arrow']}
      , {
        // className={DropdownStyles['sbui-popover__arrow']}
        offset: 10
      }, void 0), header && jsxRuntime.jsx("div", Object.assign({
        className: __styles.header
      }, {
        children: header
      }), void 0), overlay, footer && jsxRuntime.jsx("div", Object.assign({
        className: __styles.footer
      }, {
        children: footer
      }), void 0)]
    }), void 0)]
  }), void 0);
}

function Close() {
  const __styles = styleHandler['default']('popover');

  return jsxRuntime.jsx(RadixPopover__namespace.Close, Object.assign({
    className: __styles.close
  }, {
    children: jsxRuntime.jsx(IconX['default'], {
      size: 14,
      strokeWidth: 2
    }, void 0)
  }), void 0);
}

function Seperator() {
  const __styles = styleHandler['default']('popover');

  return jsxRuntime.jsx("div", {
    className: __styles.seperator
  }, void 0);
}

Popover.Seperator = Seperator;
Popover.Close = Close;

exports['default'] = Popover;
