// @supabase/postgres-meta/dist/main/PostgresMeta.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const PostgresMetaColumns_1 = __importDefault(require("./PostgresMetaColumns"));
const PostgresMetaConfig_1 = __importDefault(require("./PostgresMetaConfig"));
const PostgresMetaExtensions_1 = __importDefault(require("./PostgresMetaExtensions"));
const PostgresMetaFunctions_1 = __importDefault(require("./PostgresMetaFunctions"));
const PostgresMetaPolicies_1 = __importDefault(require("./PostgresMetaPolicies"));
const PostgresMetaPublications_1 = __importDefault(require("./PostgresMetaPublications"));
const PostgresMetaRoles_1 = __importDefault(require("./PostgresMetaRoles"));
const PostgresMetaSchemas_1 = __importDefault(require("./PostgresMetaSchemas"));
const PostgresMetaTables_1 = __importDefault(require("./PostgresMetaTables"));
const PostgresMetaTriggers_1 = __importDefault(require("./PostgresMetaTriggers"));
const PostgresMetaTypes_1 = __importDefault(require("./PostgresMetaTypes"));
const PostgresMetaVersion_1 = __importDefault(require("./PostgresMetaVersion"));
const db_1 = require("./db");
class PostgresMeta {
    constructor(config) {
        const { query, end } = db_1.init(config);
        this.query = query;
        this.end = end;
        this.columns = new PostgresMetaColumns_1.default(this.query);
        this.config = new PostgresMetaConfig_1.default(this.query);
        this.extensions = new PostgresMetaExtensions_1.default(this.query);
        this.functions = new PostgresMetaFunctions_1.default(this.query);
        this.policies = new PostgresMetaPolicies_1.default(this.query);
        this.publications = new PostgresMetaPublications_1.default(this.query);
        this.roles = new PostgresMetaRoles_1.default(this.query);
        this.schemas = new PostgresMetaSchemas_1.default(this.query);
        this.tables = new PostgresMetaTables_1.default(this.query);
        this.triggers = new PostgresMetaTriggers_1.default(this.query);
        this.types = new PostgresMetaTypes_1.default(this.query);
        this.version = new PostgresMetaVersion_1.default(this.query);
    }
}
exports.default = PostgresMeta;
