// @supabase/postgres-meta/dist/main/PostgresMetaExtensions.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const pg_format_1 = require("pg-format");
const sql_1 = require("./sql");
class PostgresMetaExtensions {
    constructor(query) {
        this.query = query;
    }
    list({ limit, offset, } = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            let sql = sql_1.extensionsSql;
            if (limit) {
                sql = `${sql} LIMIT ${limit}`;
            }
            if (offset) {
                sql = `${sql} OFFSET ${offset}`;
            }
            return yield this.query(sql);
        });
    }
    retrieve({ name }) {
        return __awaiter(this, void 0, void 0, function* () {
            const sql = `${sql_1.extensionsSql} WHERE name = ${pg_format_1.literal(name)};`;
            const { data, error } = yield this.query(sql);
            if (error) {
                return { data, error };
            }
            else if (data.length === 0) {
                return { data: null, error: { message: `Cannot find an extension named ${name}` } };
            }
            else {
                return { data: data[0], error };
            }
        });
    }
    create({ name, schema, version, cascade = false, }) {
        return __awaiter(this, void 0, void 0, function* () {
            const sql = `
CREATE EXTENSION ${pg_format_1.ident(name)}
  ${schema === undefined ? '' : `SCHEMA ${pg_format_1.ident(schema)}`}
  ${version === undefined ? '' : `VERSION ${pg_format_1.literal(version)}`}
  ${cascade ? 'CASCADE' : ''};`;
            const { error } = yield this.query(sql);
            if (error) {
                return { data: null, error };
            }
            return yield this.retrieve({ name });
        });
    }
    update(name, { update = false, version, schema, }) {
        return __awaiter(this, void 0, void 0, function* () {
            let updateSql = '';
            if (update) {
                updateSql = `ALTER EXTENSION ${pg_format_1.ident(name)} UPDATE ${version === undefined ? '' : `TO ${pg_format_1.literal(version)}`};`;
            }
            const schemaSql = schema === undefined ? '' : `ALTER EXTENSION ${pg_format_1.ident(name)} SET SCHEMA ${pg_format_1.ident(schema)};`;
            const sql = `BEGIN; ${updateSql} ${schemaSql} COMMIT;`;
            const { error } = yield this.query(sql);
            if (error) {
                return { data: null, error };
            }
            return yield this.retrieve({ name });
        });
    }
    remove(name, { cascade = false } = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const { data: extension, error } = yield this.retrieve({ name });
            if (error) {
                return { data: null, error };
            }
            const sql = `DROP EXTENSION ${pg_format_1.ident(name)} ${cascade ? 'CASCADE' : 'RESTRICT'};`;
            {
                const { error } = yield this.query(sql);
                if (error) {
                    return { data: null, error };
                }
            }
            return { data: extension, error: null };
        });
    }
}
exports.default = PostgresMetaExtensions;
