// @supabase/postgres-meta/dist/main/PostgresMetaSchemas.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const pg_format_1 = require("pg-format");
const constants_1 = require("./constants");
const sql_1 = require("./sql");
class PostgresMetaSchemas {
    constructor(query) {
        this.query = query;
    }
    list({ includeSystemSchemas = false, limit, offset, } = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            let sql = sql_1.schemasSql;
            if (!includeSystemSchemas) {
                sql = `${sql} AND NOT (n.nspname IN (${constants_1.DEFAULT_SYSTEM_SCHEMAS.map(pg_format_1.literal).join(',')}))`;
            }
            if (limit) {
                sql = `${sql} LIMIT ${limit}`;
            }
            if (offset) {
                sql = `${sql} OFFSET ${offset}`;
            }
            return yield this.query(sql);
        });
    }
    retrieve({ id, name, }) {
        return __awaiter(this, void 0, void 0, function* () {
            if (id) {
                const sql = `${sql_1.schemasSql} AND n.oid = ${pg_format_1.literal(id)};`;
                const { data, error } = yield this.query(sql);
                if (error) {
                    return { data, error };
                }
                else if (data.length === 0) {
                    return { data: null, error: { message: `Cannot find a schema with ID ${id}` } };
                }
                else {
                    return { data: data[0], error };
                }
            }
            else if (name) {
                const sql = `${sql_1.schemasSql} AND n.nspname = ${pg_format_1.literal(name)};`;
                const { data, error } = yield this.query(sql);
                if (error) {
                    return { data, error };
                }
                else if (data.length === 0) {
                    return { data: null, error: { message: `Cannot find a schema named ${name}` } };
                }
                else {
                    return { data: data[0], error };
                }
            }
            else {
                return { data: null, error: { message: 'Invalid parameters on schema retrieve' } };
            }
        });
    }
    create({ name, owner = 'postgres', }) {
        return __awaiter(this, void 0, void 0, function* () {
            const sql = `CREATE SCHEMA ${pg_format_1.ident(name)} AUTHORIZATION ${pg_format_1.ident(owner)};`;
            const { error } = yield this.query(sql);
            if (error) {
                return { data: null, error };
            }
            return yield this.retrieve({ name });
        });
    }
    update(id, { name, owner }) {
        return __awaiter(this, void 0, void 0, function* () {
            const { data: old, error } = yield this.retrieve({ id });
            if (error) {
                return { data: null, error };
            }
            const nameSql = name === undefined ? '' : `ALTER SCHEMA ${pg_format_1.ident(old.name)} RENAME TO ${pg_format_1.ident(name)};`;
            const ownerSql = owner === undefined ? '' : `ALTER SCHEMA ${pg_format_1.ident(old.name)} OWNER TO ${pg_format_1.ident(owner)};`;
            const sql = `BEGIN; ${ownerSql} ${nameSql} COMMIT;`;
            {
                const { error } = yield this.query(sql);
                if (error) {
                    return { data: null, error };
                }
            }
            return yield this.retrieve({ id });
        });
    }
    remove(id, { cascade = false } = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const { data: schema, error } = yield this.retrieve({ id });
            if (error) {
                return { data: null, error };
            }
            const sql = `DROP SCHEMA ${pg_format_1.ident(schema.name)} ${cascade ? 'CASCADE' : 'RESTRICT'};`;
            {
                const { error } = yield this.query(sql);
                if (error) {
                    return { data: null, error };
                }
            }
            return { data: schema, error: null };
        });
    }
}
exports.default = PostgresMetaSchemas;
