// @supabase/shared-types/out/projects.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackupSchedulePayloadType = exports.InitializationType = exports.InitializationPayloadVersion = exports.ProjectTask = void 0;
var ProjectTask;
(function (ProjectTask) {
    ProjectTask["UpdateJwt"] = "update_jwt";
    ProjectTask["UpdateSupabaseApiKey"] = "update_supabase_api_key";
})(ProjectTask = exports.ProjectTask || (exports.ProjectTask = {}));
var InitializationPayloadVersion;
(function (InitializationPayloadVersion) {
    InitializationPayloadVersion[InitializationPayloadVersion["Version1"] = 1] = "Version1";
})(InitializationPayloadVersion = exports.InitializationPayloadVersion || (exports.InitializationPayloadVersion = {}));
var InitializationType;
(function (InitializationType) {
    InitializationType["NewProject"] = "new_project";
    InitializationType["Restore"] = "restore";
})(InitializationType = exports.InitializationType || (exports.InitializationType = {}));
var BackupSchedulePayloadType;
(function (BackupSchedulePayloadType) {
    BackupSchedulePayloadType["Create"] = "create";
    BackupSchedulePayloadType["Update"] = "update";
    BackupSchedulePayloadType["Delete"] = "delete";
})(BackupSchedulePayloadType = exports.BackupSchedulePayloadType || (exports.BackupSchedulePayloadType = {}));
