// @supabase/ui/dist/esm/components/SidePanel/SidePanel2.js
import { __rest } from '../../_virtual/_tslib.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import * as Dialog from '@radix-ui/react-dialog';
import styleHandler from '../../lib/theme/styleHandler.js';
import Button from '../Button/Button2.js';

const SidePanel = _a => {
  var {
    id,
    className,
    children,
    header,
    visible,
    open,
    size = 'medium',
    loading,
    align = 'right',
    hideFooter = false,
    customFooter = undefined,
    onConfirm,
    onCancel,
    confirmText = 'Confirm',
    cancelText = 'Cancel',
    triggerElement,
    defaultOpen
  } = _a,
      props = __rest(_a, ["id", "className", "children", "header", "visible", "open", "size", "loading", "align", "hideFooter", "customFooter", "onConfirm", "onCancel", "confirmText", "cancelText", "triggerElement", "defaultOpen"]);

  const __styles = styleHandler('sidepanel'); // function stopPropagation(e: React.MouseEvent) {
  //   e.stopPropagation()
  // }


  const footerContent = customFooter ? customFooter : jsxs("div", Object.assign({
    className: __styles.footer
  }, {
    children: [jsx(Button, Object.assign({
      disabled: loading,
      type: "default",
      onClick: () => onCancel ? onCancel() : null
    }, {
      children: cancelText
    }), void 0), jsx(Button, Object.assign({
      loading: loading,
      onClick: () => onConfirm ? onConfirm() : null
    }, {
      children: confirmText
    }), void 0)]
  }), void 0);

  function handleOpenChange(open) {
    if (visible !== undefined && !open) {
      // controlled component behaviour
      if (onCancel) onCancel();
    }
  }

  open = open || visible;
  return jsxs(Dialog.Root, Object.assign({
    open: open,
    onOpenChange: handleOpenChange,
    defaultOpen: defaultOpen
  }, {
    children: [triggerElement && jsx(Dialog.Trigger, Object.assign({
      className: __styles.trigger
    }, {
      children: triggerElement
    }), void 0), jsxs(Dialog.Portal, {
      children: [jsx(Dialog.Overlay, {
        className: __styles.overlay
      }, void 0), jsxs(Dialog.Content, Object.assign({
        className: [__styles.base, __styles.size[size], __styles.align[align], className && className].join(' '),
        onOpenAutoFocus: props.onOpenAutoFocus,
        onCloseAutoFocus: props.onCloseAutoFocus,
        onEscapeKeyDown: props.onEscapeKeyDown,
        onPointerDownOutside: props.onPointerDownOutside,
        onInteractOutside: props.onInteractOutside
      }, {
        children: [header && jsx("header", Object.assign({
          className: __styles.header
        }, {
          children: header
        }), void 0), jsx("div", Object.assign({
          className: __styles.contents
        }, {
          children: children
        }), void 0), !hideFooter && footerContent]
      }), void 0)]
    }, void 0)]
  }), void 0);
};

function Seperator() {
  let __styles = styleHandler('sidepanel');

  return jsx("div", {
    className: __styles.seperator
  }, void 0);
}
function Content({
  children
}) {
  let __styles = styleHandler('sidepanel');

  return jsx("div", Object.assign({
    className: __styles.content
  }, {
    children: children
  }), void 0);
}
SidePanel.Content = Content;
SidePanel.Seperator = Seperator;

export { Content, Seperator, SidePanel as default };
