// @supabase/ui/dist/cjs/components/Alert/Alert2.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var IconAlertTriangle = require('../Icon/icons/IconAlertTriangle/IconAlertTriangle2.js');
var IconInfo = require('../Icon/icons/IconInfo/IconInfo2.js');
var IconX = require('../Icon/icons/IconX/IconX2.js');
var styleHandler = require('../../lib/theme/styleHandler.js');
var IconAlertOctagon = require('../Icon/icons/IconAlertOctagon/IconAlertOctagon2.js');
var IconCheckCircle = require('../Icon/icons/IconCheckCircle/IconCheckCircle2.js');

const icons = {
  danger: jsxRuntime.jsx(IconAlertOctagon['default'], {
    strokeWidth: 1.5,
    size: 18
  }, void 0),
  success: jsxRuntime.jsx(IconCheckCircle['default'], {
    strokeWidth: 1.5,
    size: 18
  }, void 0),
  warning: jsxRuntime.jsx(IconAlertTriangle['default'], {
    strokeWidth: 1.5,
    size: 18
  }, void 0),
  info: jsxRuntime.jsx(IconInfo['default'], {
    strokeWidth: 1.5,
    size: 18
  }, void 0),
  neutral: jsxRuntime.jsx(jsxRuntime.Fragment, {}, void 0)
};

function Alert({
  variant = 'neutral',
  className,
  title,
  withIcon,
  closable,
  children,
  icon,
  actions
}) {
  let __styles = styleHandler['default']('alert');

  const [visible, setVisible] = React.useState(true);
  let containerClasses = [__styles.base];
  containerClasses.push(__styles.variant[variant].base);
  if (className) containerClasses.push(className);
  let descriptionClasses = [__styles.description, __styles.variant[variant].description];
  let closeButtonClasses = [__styles.close];
  return jsxRuntime.jsx(jsxRuntime.Fragment, {
    children: visible && jsxRuntime.jsxs("div", Object.assign({
      className: containerClasses.join(' ')
    }, {
      children: [withIcon ? jsxRuntime.jsx("div", Object.assign({
        className: __styles.variant[variant].icon
      }, {
        children: withIcon && icons[variant]
      }), void 0) : null, icon && icon, jsxRuntime.jsxs("div", Object.assign({
        className: "flex flex-1 items-center justify-between"
      }, {
        children: [jsxRuntime.jsxs("div", {
          children: [jsxRuntime.jsx("h3", Object.assign({
            className: [__styles.variant[variant].header, __styles.header].join(' ')
          }, {
            children: title
          }), void 0), jsxRuntime.jsx("div", Object.assign({
            className: descriptionClasses.join(' ')
          }, {
            children: children
          }), void 0)]
        }, void 0), actions]
      }), void 0), closable && jsxRuntime.jsx("button", Object.assign({
        "aria-label": "Close alert",
        onClick: () => setVisible(false),
        className: closeButtonClasses.join(' ')
      }, {
        children: jsxRuntime.jsx(IconX['default'], {
          strokeWidth: 2,
          size: 16
        }, void 0)
      }), void 0)]
    }), void 0)
  }, void 0);
}

exports['default'] = Alert;
